File: packages/agents/src/post-production/master-editor.ts
// ---------------------------------------------------------------------------
// MasterEditorAgent — concatenates scene video clips into a final master edit
// Port of Python src/agents/master_editor.py
// ---------------------------------------------------------------------------

import { ProductionPhase } from "@marionette/shared"
import { BaseAgent } from "../base/agent.js"
import type { AgentInput, AgentOutput } from "../base/agent.js"
import { resolve, join } from "node:path"
import { mkdir, readdir, writeFile, unlink, stat } from "node:fs/promises"

// ─── Constants ───

/** Minimum file size (bytes) to consider a real video file, not a stub. */
const MIN_VIDEO_SIZE = 10_240 // 10 KB

// ─── Input extension ───

export interface MasterEditorInput extends AgentInput {
  inputDir?: string
  outputDir?: string
}

// ─── Agent ───

export class MasterEditorAgent extends BaseAgent {
  readonly name = "MasterEditor"
  readonly phase = ProductionPhase.POST
  readonly description =
    "Concatenates scene video clips into a single master edit using ffmpeg"

  async execute(input: MasterEditorInput): Promise<AgentOutput> {
    const { projectId, runId } = input
    const inputDir = resolve(input.inputDir ?? "output/videos")
    const outputDir = resolve(input.outputDir ?? "output/master")

    await mkdir(outputDir, { recursive: true })

    this.log("Starting ffmpeg video concatenation process...")
    await this.updateProgress(runId, 5)

    // 1. Find .mp4 files in input directory that are large enough to be real
    const videoFiles = await this.findVideoFiles(inputDir)

    if (videoFiles.length === 0) {
      this.log("No valid video files found, creating mock master edit")
      const mockPath = await this.createMockMaster(outputDir, projectId)

      await this.saveAsset({
        projectId,
        type: "VIDEO",
        agentName: this.name,
        filePath: mockPath,
        fileName: mockPath.split("/").pop() ?? "master_edit_mock.mp4",
        mimeType: "text/plain",
        metadata: { mock: true },
      })

      await this.updateProgress(runId, 100)

      return {
        success: true,
        message:
          "No real video files found. Created mock master edit placeholder.",
        outputPath: mockPath,
        data: { mock: true },
      }
    }

    this.log(`Found ${videoFiles.length} video clips to merge`)
    await this.updateProgress(runId, 20)

    // 2. Create concat.txt manifest for ffmpeg
    const concatFilePath = join(outputDir, "concat.txt")
    const concatContent = videoFiles
      .map((filePath) => `file '${filePath}'`)
      .join("\n")
    await writeFile(concatFilePath, concatContent, "utf-8")

    // 3. Run ffmpeg
    const masterPath = join(outputDir, `final_master_${projectId}.mp4`)

    try {
      await this.updateProgress(runId, 40)
      this.log("Merging video clips with ffmpeg...")

      const result = await Bun.spawn({
        cmd: [
          "ffmpeg",
          "-y",
          "-f",
          "concat",
          "-safe",
          "0",
          "-i",
          concatFilePath,
          "-c:v",
          "libx264",
          "-preset",
          "fast",
          "-crf",
          "23",
          "-c:a",
          "aac",
          "-movflags",
          "+faststart",
          masterPath,
        ],
        env: {
          ...process.env,
          PATH: `/opt/homebrew/bin:${process.env["HOME"]}/.bun/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:${process.env["PATH"] ?? ""}`,
        },
        stdout: "ignore",
        stderr: "pipe",
      }).exited

      if (result !== 0) {
        throw new Error(`ffmpeg exited with code ${result}`)
      }

      this.log(`Master edit rendered: ${masterPath}`)

      // Save asset to DB
      const fileStats = await stat(masterPath).catch(() => null)
      await this.saveAsset({
        projectId,
        type: "VIDEO",
        agentName: this.name,
        filePath: masterPath,
        fileName: masterPath.split("/").pop() ?? "final_master.mp4",
        mimeType: "video/mp4",
        fileSize: fileStats?.size,
        metadata: {
          clipCount: videoFiles.length,
          inputDir,
        },
      })

      await this.updateProgress(runId, 100)

      return {
        success: true,
        message: `Master edit complete: merged ${videoFiles.length} clips`,
        outputPath: masterPath,
        data: {
          clipCount: videoFiles.length,
          masterPath,
        },
      }
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : String(err)
      this.log(`ffmpeg error: ${errorMessage}`)
      this.log(
        "Ensure ffmpeg is installed (brew install ffmpeg)",
      )

      return {
        success: false,
        message: `ffmpeg concatenation failed: ${errorMessage}`,
        data: { clipCount: videoFiles.length },
      }
    } finally {
      // Clean up concat.txt
      await unlink(concatFilePath).catch(() => {
        // Ignore cleanup errors (sandbox restrictions, etc.)
      })
    }
  }

  // ── Helpers ────────────────────────────────────────────────────

  private async findVideoFiles(dir: string): Promise<string[]> {
    let entries: string[]
    try {
      entries = await readdir(dir)
    } catch {
      this.log(`Input directory not found: ${dir}`)
      return []
    }

    const mp4Files = entries
      .filter((name) => name.endsWith(".mp4"))
      .sort()

    const validFiles: string[] = []
    for (const name of mp4Files) {
      const filePath = join(dir, name)
      const fileStats = await stat(filePath).catch(() => null)
      if (fileStats && fileStats.size > MIN_VIDEO_SIZE) {
        validFiles.push(filePath)
      }
    }

    return validFiles
  }

  private async createMockMaster(
    outputDir: string,
    projectId: string,
  ): Promise<string> {
    const filename = `master_edit_${projectId}_mock.mp4`
    const filePath = join(outputDir, filename)
    const content =
      `Mock Master Edit File for project: '${projectId}'.\n` +
      `Please add real MP4 files to the videos directory.\n`
    await writeFile(filePath, content, "utf-8")
    return filePath
  }
}

File: apps/production-pipeline/src/agents/master_editor.py
import os
from typing import Optional, Union, List
import json
import glob
import subprocess

class MasterEditorAgent:
    def __init__(self, output_dir: str = "output/master", videos_dir: str = "output/videos"):
        """
        AI 마스터 에디터 초기화
        """
        self.output_dir = output_dir
        self.videos_dir = videos_dir
        os.makedirs(self.output_dir, exist_ok=True)
        
    def edit_and_color_grade(self, json_path: str):
        """
        ouput/videos 폴더 내의 영상을 하나로 연결(Concatenate)합니다.
        """
        print(f"✂️ AI 마스터 에디터: FFMPEG 영상 병합(Concatenation) 프로세스를 시작합니다...")
        
        with open(json_path, "r", encoding="utf-8") as f:
            data = json.load(f)
            title = data.get("title", "Untitled").replace(" ", "_")
            
        # 1. 대상 비디오 파일 탐색 (확장자 mp4, 최소 10KB 이상인 실제 비디오만)
        all_mp4 = sorted(glob.glob(os.path.join(self.videos_dir, "*.mp4")))
        video_files = [v for v in all_mp4 if os.path.getsize(v) > 10240]  # 10KB 이상

        if not video_files:
            print("⚠️ 병합할 영상 파일(mp4)이 없으므로 더미 모드로 처리합니다.")
            master_filename = f"master_edit_{title}.mp4"
            master_path = os.path.join(self.output_dir, master_filename)
            with open(master_path, "w", encoding="utf-8") as master_file:
                master_file.write(f"Mock Master Edit File for: '{title}'. Please add real MP4s to output/videos.\n")
            return master_path

        print(f"🔍 병합 대상 클립 {len(video_files)}개 발견. FFMPEG 콘캣(Concat)을 준비합니다.")
        
        # 2. FFMPEG 리스트 파일(concat.txt) 생성
        list_file_path = os.path.join(self.output_dir, "concat.txt")
        with open(list_file_path, "w", encoding="utf-8") as list_file:
            for vid in video_files:
                # ffmpeg는 concat 파일 안의 경로를 안전하게 읽기 위해 절대 경로 또는 형식 필요
                abs_path = os.path.abspath(vid)
                list_file.write(f"file '{abs_path}'\n")
                
        # 3. FFMPEG 실행 (무손실 코덱 복사 방식)
        master_path = os.path.join(self.output_dir, f"final_master_{title}.mp4")
        
        # 기존 파일 덮어쓰기 로직
        if os.path.exists(master_path):
            os.remove(master_path)
            
        ffmpeg_cmd = [
            "ffmpeg", "-y", "-f", "concat", "-safe", "0",
            "-i", list_file_path,
            "-c", "copy",
            master_path
        ]
        
        try:
            print(f"🎥 영상을 연결하고 있습니다... (ffmpeg)")
            subprocess.run(ffmpeg_cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            print(f"   ✅ 영상 컷 편집/병합 렌더링 완료: {master_path}")
        except subprocess.CalledProcessError as e:
            print(f"❌ FFMPEG 실행 중 오류가 발생했습니다. (ffmpeg가 설치되어 있는지 확인하세요: brew install ffmpeg)")
            print(f"에러 코드: {e}")
            return None
        finally:
            # 임시 파일 정리
            try:
                if os.path.exists(list_file_path):
                    os.remove(list_file_path)
            except PermissionError:
                pass  # 샌드박스 환경에서는 삭제 불가할 수 있음

        print(f"🎉 모든 영상 마스터 편집 병합 완료!")
        return master_path
